components/safety_board/src/SYS_Vehicle.c

```diff
@@ -0,0 +1,255 @@
+    {
+        HAL_GPIO_WritePin(PCHG_Port, PCHG_Pin, (state == ON) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+    }
+    else if (contactor == MAIN_CONTACTOR)
+    {
+        HAL_GPIO_WritePin(AIR_Port, AIR_Pin, (state == ON) ? GPIO_PIN_SET : GPIO_PIN_RESET);
+    }
+}
+
+void SYS_SAFETY_SetBatteryVoltage(int16_t v)
+{
+    batt_voltage      = v;
+    last_batt_message = HAL_GetTick();
+    SYS_SAFETY_SetStatus(BMS_STATUS, ON);
+}
+
+void SYS_SAFETY_SetBusVoltage(int16_t v)
+{
+    ts_voltage      = v;
+    last_ts_message = HAL_GetTick();
+    SYS_SAFETY_SetStatus(TS_STATUS, ON);
+}
+
+void SYS_SAFETY_SetIsolation(uint8_t duty, uint8_t freq)
+{
+    imd_duty_cycle = duty;
+    imd_freq = freq;    
+}
+
+void SYS_SAFETY_CycleState(void)
+{
+    SYS_Validate();
+
+    if ((status_flags & (0x01 << IMD_STATUS)) && (status_flags & (0x01 << BMS_STATUS)) && (vehicle_state == VEHICLE_FAULT))
+    {
+            vehicle_state = VEHICLE_READY;
+    } 
+    
+    if ((vehicle_state == VEHICLE_READY) && (((uint16_t) status_flags & (0x01 << TSMS_STATUS)) != 0))
+    {
+        SYS_CONTACTORS_Switch(PRECHARGE_CONTACTOR, ON);
+        vehicle_state = CONTACTOR_PRECHARGE_CLOSED;
+    } 
+    else if (vehicle_state == CONTACTOR_PRECHARGE_CLOSED)
+    {
+        if (ts_voltage > ((batt_voltage * 95) / 100))
+        {
+            vehicle_state = CONTACTOR_PRECHARGE_MAIN_CLOSED;
+            SYS_CONTACTORS_Switch(MAIN_CONTACTOR, ON);
+        }
+        else if ((status_flags & (0x01 << BMS_CHARGE_STATUS)) && !(status_flags & 0x01 << BMS_DISCHARGE_STATUS))
+        {
+            // TODO: Complete charger heartbeat validation
+            SYS_CONTACTORS_Switch(MAIN_CONTACTOR, ON);
+            vehicle_state = CONTACTOR_MAIN_CLOSED;
+        }
+    }
+    else if (vehicle_state == CONTACTOR_PRECHARGE_MAIN_CLOSED)
+    {
+        if (ts_voltage > ((batt_voltage * 98) / 100))
+        {
+            SYS_CONTACTORS_Switch(MAIN_CONTACTOR, ON);
+            vehicle_state = CONTACTOR_MAIN_CLOSED;
+        }
+    }
+
+    CAN_data_T data;
+
+    data.u8[0] = status_flags;
+    data.u8[1] = vehicle_state;
+    data.u8[2] = batt_voltage & 0xff;
+    data.u8[3] = (batt_voltage & 0xff00) >> 8;
+    data.u8[4] = ts_voltage & 0xff;
+    data.u8[5] = (ts_voltage & 0xff00) >> 8;
+    data.u8[6] = imd_duty_cycle;
+    data.u8[7] = imd_freq;
+    
+    extern CAN_HandleTypeDef hcan;
+    static uint8_t rst_counter = 0;
+
+    if (!CAN_sendMsgBus0(CAN_TX_PRIO_100HZ, data, SAFETY_BOARD_STATUS_ID, 8)
+        && rst_counter++ == 100)
+    {
+        HAL_NVIC_SystemReset();
+    }
+    else if (HAL_CAN_GetError(&hcan))
+    {
+        HAL_CAN_ResetError(&hcan);
+    }
+    else 
+    {
+        rst_counter = 0;
+    }
+
+}
+
+/******************************************************************************
+ *                     P R I V A T E  F U N C T I O N S
+ ******************************************************************************/
+
+bool SYS_Validate(void)
```

Comment:
I think all this logic needs to be broken up into at least one state machine. I can help define it